llm_demo/orchestrator/langgraph/langgraph_orchestrator.py

```diff
@@ -46,131 +49,125 @@
-        model: str,
-    ) -> "UserAgent":
-        llm = VertexAI(max_output_tokens=512, model_name=model, temperature=0.0)
-        memory = ConversationBufferMemory(
-            chat_memory=ChatMessageHistory(messages=history),
-            memory_key="chat_history",
-            input_key="input",
-            output_key="output",
-        )
-        agent = initialize_agent(
-            tools,
-            llm,
-            agent=AgentType.STRUCTURED_CHAT_ZERO_SHOT_REACT_DESCRIPTION,
-            memory=memory,
-            handle_parsing_errors=True,
-            max_iterations=3,
-            early_stopping_method="generate",
-            return_intermediate_steps=True,
-        )
-        agent.agent.llm_chain.prompt = prompt  # type: ignore
-        return UserAgent(client, agent, memory)
-
-    async def close(self):
-        await self.client.close()
-
-    async def invoke(self, prompt: str) -> Dict[str, Any]:
-        try:
-            response = await self.agent.ainvoke({"input": prompt})
-        except Exception as err:
-            raise HTTPException(status_code=500, detail=f"Error invoking agent: {err}")
-        return response
-
-    async def insert_ticket(self, params: str):
-        return await insert_ticket(self.client, params)
-
-    def reset_memory(self, base_message: List[BaseMessage]):
-        self.memory.clear()
-        self.memory.chat_memory = ChatMessageHistory(messages=base_message)
+    is_last_step: IsLastStep
 
 
 class LangGraphOrchestrator(BaseOrchestrator):
-    _user_sessions: Dict[str, UserAgent]
+    _user_sessions: List[str]
     # aiohttp context
     connector = None
 
     def __init__(self):
-        self._user_sessions = {}
+        self._user_sessions = []
 
     @classproperty
     def kind(cls):
         return "langgraph"
 
+    async def create_graph(self, tools, prompt: ChatPromptTemplate, model_name: str):
+        # tool node
+        tool_node = ToolNode(tools)
+
+        # model node
+        model = ChatVertexAI(
+            model=model_name, temperature=0, max_tokens=512, max_iterations=3
+        ).bind_tools(tools)
+        # Add the prompt
+        model_runnable = (lambda messages: prompt + messages) | model
+
+        def call_model(state: UserState, config: RunnableConfig):
+            """
+            Node for function that calls the model
+            """
+            messages = state["messages"]
+            response = model_runnable.invoke(messages, config)
+            if (
+                state["is_last_step"]
+                and isinstance(response, AIMessage)
+                and response.tool_calls
+            ):
+                return {
+                    "messages": [
+                        AIMessage(
+                            id=response.id,
+                            content="Sorry, need more steps to process this request.",
+                        )
+                    ]
+                }
+            return {"messages": [response]}
```

Comment:
Why is this function inline? Can we get a comment or explanation as to what it is doing?

Reply:
This is written similarly to the prebuilt react_agent. Let me know if it's better to move it out (so that this isn't inline).